plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +229,127 @@ class bp_connection_manager {
+            return !host.empty() && !port.empty();
+         };
+         std::map<eosio::name, size_t> producers;
+         const gossip_bp_peers_message::bp_peer* prev = nullptr;
+         for (const auto& peer : msg.peers) {
+            if (peer.producer_name.empty())
+               return false;
+            if (!valid_address(peer.server_address))
+               return false;
+            if (prev != nullptr) {
+               if (prev->producer_name == peer.producer_name) {
+                  if (prev->server_address == peer.server_address)
+                     return false; // duplicate entries not allowed
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               }
+            }
+            if (++producers[peer.producer_name] > 4)
+               return false; // only 4 entries per producer allowed
+            prev = &peer;
+         }
+      }
+
+      controller& cc = self()->chain_plug->chain();
+
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& sig_idx = gossip_bps.index.get<by_sig>();
+      for (auto i = msg.peers.begin(); i != msg.peers.end();) {
+         const auto& peer = *i;
+         try {
+            if (!sig_idx.contains(peer.sig)) { // we already have it, already verified
+               // peer key may have changed or been removed, so if invalid or not found skip it
+               if (std::optional<public_key_type> peer_key = cc.get_peer_key(peer.producer_name)) {
+                  public_key_type pk(peer.sig, peer.digest());
+                  if (pk != *peer_key) {
+                     fc_dlog(self()->get_logger(), "Recovered peer key did not match on-chain ${p}, recovered: ${pk} != expected: ${k}",
+                             ("p", peer.producer_name)("pk", pk)("k", *peer_key));
+                     i = msg.peers.erase(i);
+                     continue;
+                  }
+               } else { // unknown key
+                  fc_dlog(self()->get_logger(), "Failed to find peer key ${p}", ("p", peer.producer_name));
+                  i = msg.peers.erase(i);
+                  continue;
+               }
+            }
+         } catch (fc::exception& e) {
+            fc_dlog(self()->get_logger(), "Exception recovering peer key ${p}, error: ${e}", ("p", peer.producer_name)("e", e.to_detail_string()));
+            // invalid key
+            i = msg.peers.erase(i);
+            continue;
+         }
+         ++i;
+      }
+
+      return !msg.peers.empty();
+   }
+
+   // thread-safe
+   bool update_gossip_bps(const gossip_bp_peers_message& msg) {
+      // providing us with full set
+      fc::unique_lock g(gossip_bps.mtx);
+      auto& idx = gossip_bps.index.get<by_producer>();
+      bool diff = false;
+      for (const auto& peer : msg.peers) {
+         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_address))); i != idx.end()) {
```

Comment:
What if the peer changes the server address it uses? Old addresses will remain in `gossip_bps`, right?

Reply:
Yes. If they change it most likely the previous one is no longer valid in which case the connection will fail. If it remains valid then everyone will connect to both. We should recommend that if they change their server address that they also register a new key on-chain.